tokio/src/runtime/dump.rs

```diff
@@ -30,14 +30,158 @@ pub struct Task {
+pub struct BacktraceFrame {
+    ip: *mut std::ffi::c_void,
+    symbol_address: *mut std::ffi::c_void,
+    symbols: Vec<BacktraceSymbol>,
+}
+
+impl BacktraceFrame {
+    pub(crate) fn from_resolved_backtrace_frame(frame: &backtrace::BacktraceFrame) -> Self {
+        Self {
+            ip: frame.ip(),
+            symbol_address: frame.symbol_address(),
+            symbols: frame
+                .symbols()
+                .iter()
+                .map(BacktraceSymbol::from_backtrace_symbol)
+                .collect(),
+        }
+    }
+
+    /// Return the instruction pointer of this frame.
+    ///
+    /// See the ABI docs for your platform for the exact meaning.
+    pub fn ip(&self) -> *mut std::ffi::c_void {
+        self.ip
+    }
+
+    /// Returns the starting symbol address of the frame of this function.
+    pub fn symbol_address(&self) -> *mut std::ffi::c_void {
+        self.symbol_address
+    }
+
+    /// Return an iterator over the symbols of this backtrace frame.
+    ///
+    /// Due to inlining, it is possible for there to be multiple [BacktraceSymbol] items relating
+    /// to a single frame. The first symbol listed is the "innermost function",
+    /// whereas the last symbol is the outermost (last caller).
+    pub fn symbols(&self) -> impl Iterator<Item = &BacktraceSymbol> {
+        self.symbols.iter()
+    }
+}
+
 /// An execution trace of a task's last poll.
 ///
+/// <div class="warning">
+/// Resolving a backtrace, either via the [`Display`][std::fmt::Display] impl or via
+/// [`resolve_backtraces`][Trace::resolve_backtraces], parses debuginfo, which is
+/// possibly a CPU-expensive operation that can take a platform-specific but
+/// long time to run - often over 100 milliseconds, especially if the current
+/// process's binary is big. In some cases, the platform might internally cache some of the
+/// debuginfo, so successive calls to `resolve_backtraces` might be faster than
+/// the first call, but all guarantees are platform-dependent.
+///
+/// To avoid blocking the runtime, it is recommended
+/// that you resolve backtraces inside of a [spawn_blocking()][crate::task::spawn_blocking]
+/// and to have some concurrency-limiting mechanism to avoid unexpected performance impact.
+/// </div>
+///
 /// See [Handle::dump][crate::runtime::Handle::dump].
 #[derive(Debug)]
 pub struct Trace {
     inner: super::task::trace::Trace,
 }
 
+impl Trace {
+    /// Resolve and return a list of backtraces that are involved in polls in this task.
+    pub fn resolve_backtraces(&self) -> Vec<Vec<BacktraceFrame>> {
+        self.inner
+            .backtraces()
+            .iter()
+            .map(|backtrace| {
+                let mut backtrace = backtrace::Backtrace::from(backtrace.clone());
+                backtrace.resolve();
+                backtrace
+                    .frames()
+                    .iter()
+                    .map(BacktraceFrame::from_resolved_backtrace_frame)
+                    .collect()
```

Comment:
Perhaps we can fix this in the follow up PR, but if we are going to have our own `Backtrace` type, I think we can avoid this conversion by making [`trace_leaf`](https://github.com/tokio-rs/tokio/blob/d08578fc9a950c8ee0ef242dbce0fcaaed5e853f/tokio/src/runtime/task/trace/mod.rs#L153) function construct tokio's Backtrace struct directly.

Reply:
I'll rather have `trace_leaf` work with the `backtrace` crate and keep "tokio Backtrace" for return purposes.